x/storage/keeper/bucket_rate_limit.go

```diff
@@ -0,0 +1,290 @@
+	internalBucketInfo.PriceTime = ctx.BlockTime().Unix()
+	bill, err := k.GetBucketReadStoreBill(ctx, bucketInfo, internalBucketInfo)
+	if err != nil {
+		return fmt.Errorf("get bucket bill failed: %s %s", bucketInfo.BucketName, err.Error())
+	}
+	err = k.paymentKeeper.ApplyUserFlowsList(ctx, []paymenttypes.UserFlows{bill})
+	if err != nil {
+		return fmt.Errorf("apply user flows list failed: %s %w", bucketInfo.BucketName, err)
+	}
+	k.SetInternalBucketInfo(ctx, bucketInfo.Id, internalBucketInfo)
+	return nil
+}
+
+func getTotalOutFlowRate(flows []paymenttypes.OutFlow) sdkmath.Int {
+	totalFlowRate := sdkmath.ZeroInt()
+	for _, flow := range flows {
+		totalFlowRate = totalFlowRate.Add(flow.Rate)
+	}
+	return totalFlowRate
+}
+
+// unChargeAndLimitBucket uncharges the bucket and limits the bucket
+func (k Keeper) unChargeAndLimitBucket(ctx sdk.Context, bucketInfo *types.BucketInfo, paymentAccount sdk.AccAddress, bucketName string) error {
+	k.setBucketFlowRateLimitStatus(ctx, bucketName, &types.BucketFlowRateLimitStatus{
+		IsBucketLimited: true,
+		PaymentAddress:  paymentAccount.String(),
+	})
+
+	internalBucketInfo := k.MustGetInternalBucketInfo(ctx, bucketInfo.Id)
+	// if the rate limit is not set and the payment account is not owned by the bucket owner,
+	// the net flow rate of the bucket should be zero, but it's fine to call `unChargeBucketReadStoreFee`, it will do nothing
+	return k.unChargeBucketReadStoreFee(ctx, bucketInfo, internalBucketInfo)
+}
+
+// setFlowRateLimit sets the flow rate limit of the bucket
+func (k Keeper) setFlowRateLimit(ctx sdk.Context, bucketInfo *types.BucketInfo, paymentAccount sdk.AccAddress, bucketName string, rateLimit sdkmath.Int) error {
+	currentRateLimit, found := k.getBucketFlowRateLimit(ctx, paymentAccount, sdk.MustAccAddressFromHex(bucketInfo.Owner), bucketName)
+
+	// do nothing if the bucket flow rate limit is already the same as the new rate limit
+	if found && currentRateLimit.FlowRateLimit.Equal(rateLimit) {
+		return nil
+	}
+
+	internalBucketInfo := k.MustGetInternalBucketInfo(ctx, bucketInfo.Id)
+	isRateLimited := k.IsBucketRateLimited(ctx, bucketName)
+
+	if found && isRateLimited {
+		internalBucketInfo.PriceTime = ctx.BlockTime().Unix()
+
+		currentBill, err := k.GetBucketReadStoreBill(ctx, bucketInfo, internalBucketInfo)
+		if err != nil {
+			return fmt.Errorf("get bucket currentBill failed: %s %s", bucketInfo.BucketName, err.Error())
+		}
+		totalOutFlowRate := getTotalOutFlowRate(currentBill.Flows)
+
+		// if the total net flow rate is less than or equal to the new rate limit,
+		// resume the charge of the bucket and delete the flow rate limit status
+		if totalOutFlowRate.LTE(rateLimit) {
+			err := k.chargeBucketReadStoreFee(ctx, bucketInfo, internalBucketInfo)
+			if err != nil {
+				return fmt.Errorf("charge bucket failed: %s %s", bucketInfo.BucketName, err.Error())
+			}
+			k.deleteBucketFlowRateLimitStatus(ctx, bucketName)
+			return nil
+		}
+
+		return nil
```

Comment:
what if it is not isRateLimited before, but `totalOutFlowRate.LTE(rateLimit)`, you do not handle this case.